### PR TITLE
Add Intermediate skills - Level 9: Dataset Explorartion from a Query Image/Text

### PR DESCRIPTION
**Pull request by sooahleex**
### Summary

- Ticket no. 108906
- Add intermediate skills - Level 9: Data exploration
- Rename searcher to explorer

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added the description of my changes into [CHANGELOG](https://github.com/openvinotoolkit/datumaro/blob/develop/CHANGELOG.md).​
- [X] I have updated the [documentation](https://github.com/openvinotoolkit/datumaro/tree/develop/docs) accordingly

### License

- [ ] I submit _my code changes_ under the same [MIT License](https://github.com/openvinotoolkit/datumaro/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

File: docs/source/docs/command-reference/context_free/explorer.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+- `-q, --query` (string) - Image path or text to use as query.
+- `-topk` (int) - Number how much you want to find similar data.
+- `-p, --project` (string) - Directory of the project to operate on (default: current directory).
+- `-s, --save` (bool) - Save visualized result of similar dataset.
+
+Examples:
+- Use image query
+```bash
+datum create <...>
+datum import -f datumaro <path/to/dataset/>
```
Comment (vinnamkim) [inline docs/source/docs/command-reference/context_free/explorer.md]:
```suggestion
datum project create <...>
datum project import -f datumaro <path/to/dataset/>
```

Please fix the other parts as well (not only this file).

File: docs/source/docs/level-up/intermediate_skills/09_data_exploration.rst [hunk 2]
```diff
@@ -2,4 +2,56 @@
...
+        With the project-based CLI, we first require to create a project by
+
+        ..code-block:: bash
+            datum project create -o <path/to/project>
+
+        We now import data in to project through
+
+        ..code-block:: bash
+
+            datum import <path/to/dataset/>
```
Comment (vinnamkim) [inline docs/source/docs/level-up/intermediate_skills/09_data_exploration.rst]: Here

File: docs/source/docs/release_notes.rst [hunk 3]
```diff
@@ -12,7 +12,7 @@ New features
 - Add with_subset_dirs decorator (Add ImagenetWithSubsetDirsImporter)
 - Add CommonSemanticSegmentationWithSubsetDirsImporter
 - Add DatumaroBinary format
-- Add Searcher CLI documentation
+- Add Explorer CLI documentation
```
Comment (vinnamkim) [inline docs/source/docs/release_notes.rst]: Do not change the old history

File: tests/integration/cli/test_explore.py [hunk 4]
```diff
@@ -17,24 +17,14 @@
 from tests.utils.test_utils import run_datum as run
 
 
-class SearchTest(TestCase):
-    @skipIf(
-        platform.system() == "Darwin",
-        "Segmentation fault only occurs on MacOS: "
-        "https://github.com/openvinotoolkit/datumaro/actions/runs/4202399957/jobs/7324077250",
-    )
```
Comment (vinnamkim) [inline tests/integration/cli/test_explore.py:25]: Is it ok to remove this?

File: datumaro/plugins/explorer.py [hunk 5]
```diff
@@ -86,3 +87,9 @@ def type_check(self, item):
         if not isinstance(item.media, Image):
             raise MediaTypeError(f"Media type should be Image, Current type={type(item.media)}")
         return True
+
+    def inferenced_check(self, item):
+        for annotation in item.annotations:
+            if isinstance(annotation, HashKey) and not annotation.hash_key:
+                return True
+        return False
```
Comment (vinnamkim) [inline datumaro/plugins/explorer.py:95]: I don't know why this PR has many code changes in the release day.

File: datumaro/components/launcher.py [hunk 6]
```diff
@@ -42,7 +42,9 @@ def __iter__(self):
         for batch in take_by(self._extractor, self._batch_size):
             inputs = []
             for item in batch:
-                if not self._launcher.type_check(item):
+                if not self._launcher.type_check(item) and not self._launcher.inferenced_check(
```
Comment (vinnamkim) [inline datumaro/components/launcher.py]: Please revert this. We don't have time to change the core component in the release day.

File: datumaro/plugins/explorer.py (hunk 5 above)
Comment (vinnamkim) [inline datumaro/plugins/explorer.py]: Please revert this. We had no time for it.

File: docs/source/docs/level-up/intermediate_skills/09_data_exploration.rst (hunk 2 above)
Comment (vinnamkim) [reply docs/source/docs/level-up/intermediate_skills/09_data_exploration.rst]: Please double check by yourself.

File: datumaro/plugins/explorer.py (hunk 5 above)
Comment (sooahleex) [reply datumaro/plugins/explorer.py]: Revert this.

File: datumaro/components/launcher.py (hunk 6 above)
Comment (sooahleex) [reply datumaro/components/launcher.py]: Revert this.

File: tests/integration/cli/test_explore.py (hunk 4 above)
Comment (sooahleex) [reply tests/integration/cli/test_explore.py:25]: This is added below.

File: docs/source/docs/release_notes.rst (hunk 3 above)
Comment (sooahleex) [reply docs/source/docs/release_notes.rst]: Fix.

File: docs/source/docs/level-up/intermediate_skills/09_data_exploration.rst (hunk 2 above)
Comment (sooahleex) [reply docs/source/docs/level-up/intermediate_skills/09_data_exploration.rst]: Fix.

File: docs/source/docs/command-reference/context_free/explorer.md (hunk 1 above)
Comment (sooahleex) [reply docs/source/docs/command-reference/context_free/explorer.md]: Fix.

File: docs/source/docs/level-up/intermediate_skills/09_data_exploration.rst [hunk 7]
```diff
@@ -2,4 +2,56 @@
 Level 9: Dataset Explorartion from a Query Image/Text
 =====================================================
 
-TBD
+
+Datumaro support exploration feature to find out similar data for query among dataset. With query, the exploration result includes top-k similar data among dataset.
+Through this feature, you could figure out dataset property. You could check the visualization result of exploration using `Visualizer`.
+
+More detailed descriptions about explorer are given by :ref:`Explore`
+The Python example for the usage of explorer is described in `here <https://github.com/openvinotoolkit/datumaro/blob/develop/notebooks/07_data_explorer.ipynb>`_.
```
Comment (vinnamkim) [inline docs/source/docs/level-up/intermediate_skills/09_data_exploration.rst]: Please check this comment

https://github.com/openvinotoolkit/datumaro/pull/960#discussion_r1172038374

File: docs/source/docs/command-reference/context_free/explorer.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+The command can be applied to a dataset. And if you want to use multiple dataset as database, you could use merged dataset. The current project (`-p/--project`) is also used a context for plugins, so it can be useful for dataset paths having custom formats. When not specified, the current project's working tree is used. To save visualized result (`-s/--save`) is turned on as default. This visualized result is based on [Visualizer](https://openvinotoolkit.github.io/datumaro/docs/python-api/python-api-examples/visualizer/).
```
Comment (vinnamkim) [inline docs/source/docs/command-reference/context_free/explorer.md]: `[Visualizer](https://openvinotoolkit.github.io/datumaro/docs/python-api/python-api-examples/visualizer/).`

This link must be broken. Please see

https://github.com/openvinotoolkit/datumaro/pull/958/files#diff-df1c619067c1f80415f16b01bf9fa5ffc9106e33c6b9432118d91edfc0adb9c8

**Review by vinnamkim — APPROVED** [review]
LGTM.

File: docs/source/docs/level-up/intermediate_skills/09_data_exploration.rst [hunk 9]
```diff
@@ -2,4 +2,58 @@
...
+
+        .. code-block:: bash
+
+            datum project create -o <path/to/project>
+
+        We now import data in to project through
+
+        .. code-block:: bash
+
+            datum project import <path/to/dataset/>
```
Comment (wonjuleee) [inline docs/source/docs/level-up/intermediate_skills/09_data_exploration.rst]: 1. `<path/to/dataset/>` -> `<path/to/data>`
2. This should be `datum project import --project <path/to/project> <path/to/data>`

File: docs/source/docs/level-up/intermediate_skills/09_data_exploration.rst [hunk 10]
```diff
@@ -2,4 +2,58 @@
...
+
+        We can explore similar items for the query
+
+        .. code-block:: bash
+
+            datum explore -q QUERY -topk TOPK_NUM
```
Comment (wonjuleee) [inline docs/source/docs/level-up/intermediate_skills/09_data_exploration.rst]: isn't there any project information?

Comment (sooahleex) [reply docs/source/docs/level-up/intermediate_skills/09_data_exploration.rst]: I omitted project path due to could catch up the current project automatically. But I added this.

File: docs/source/docs/level-up/intermediate_skills/09_data_exploration.rst (hunk 9 above)
Comment (sooahleex) [reply docs/source/docs/level-up/intermediate_skills/09_data_exploration.rst]: Fix.

**Review by cih9088 — APPROVED** [review]

**Pull request merged**